mmcls/datasets/builder.py

```diff
@@ -83,6 +102,8 @@ def build_dataloader(dataset,
     """
     rank, world_size = get_dist_info()
 
+    sampler_cfg = get_sampler_cfg(cfg, distributed=dist)
```

Comment:
It is not a good choice to pass the whole config into a module's builder.

What if we change like this?

```Python
    rank, world_size = get_dist_info()
    # default logic
    if dist:
        sampler = DistributedSampler(
            dataset, world_size, rank, shuffle=shuffle, round_up=round_up)
        batch_size = samples_per_gpu
        num_workers = workers_per_gpu
    else:
        sampler = None
        batch_size = num_gpus * samples_per_gpu
        num_workers = num_gpus * workers_per_gpu

    # build custom sampler logic
    if sampler_cfg:
        # shuffle=False when val and test
        sampler_cfg.update(shuffle=shuffle)
        sampler = build_sampler(
            sampler_cfg,
            default_args=dict(dataset=dataset, num_replicas=world_size, rank=rank))

    # If sampler exists, turn off dataloader shuffle
    if sampler is not None:
        shuffle = False
```

Reply:
I think it's better to use:
```python
    rank, world_size = get_dist_info()

    # Custom sampler logic
    if sampler_cfg:
        # shuffle=False when val and test
        sampler_cfg.update(shuffle=shuffle)
        sampler = build_sampler(
            sampler_cfg,
            default_args=dict(dataset=dataset, num_replicas=world_size, rank=rank))
    # Default sampler logic
    elif dist:
        sampler = DistributedSampler(
            dataset, world_size, rank, shuffle=shuffle, round_up=round_up)
    else:
        sampler = None

    # If sampler exists, turn off dataloader shuffle
    if sampler is not None:
        shuffle = False

    if dist:
        batch_size = samples_per_gpu
        num_workers = workers_per_gpu
    else:
        batch_size = num_gpus * samples_per_gpu
        num_workers = num_gpus * workers_per_gpu
```
Then we can avoid a unnecessary instantiation of `DistributedSampler`. And split the logic of sampler and batch_size.